oauthenticator/google.py

```diff
@@ -27,6 +28,16 @@ def check_user_in_groups(member_groups, allowed_groups):
     else:
         return False
 
+class GoogleLoginHandler(OAuthLoginHandler):
+    """See https://developers.google.com/identity/protocols/oauth2 for general information."""
+
+    def authorize_redirect(self, *args, **kwargs):
+        """Add `access_type`, `approval_prompt` to redirect params"""
+        extra_params = kwargs.setdefault('extra_params', {})
+        extra_params["access_type"] = 'offline'
+        extra_params["approval_prompt"] = 'force'
```

Comment:
can you explain a bit why this change is needed and what it will do? Especially what effect will it have on people already using this authenticator.

At first glance it seems unrelated to storing the refresh token.

Reply:
These changes are to get an `access_token` and `refresh_token` for all scopes at once. In the case that people are adding more scopes to their login request, example: to give users access to Google Drive.

Reply:
I could also work on making these something people opt-in

Reply:
maybe something like this to make it something people opt-in

```diff
diff --git a/oauthenticator/google.py b/oauthenticator/google.py
index b5b1e31..7fc3bf3 100644
--- a/oauthenticator/google.py
+++ b/oauthenticator/google.py
@@ -13,7 +13,7 @@ from tornado.httpclient import HTTPRequest, AsyncHTTPClient
 from tornado.auth import GoogleOAuth2Mixin
 from tornado.web import HTTPError

-from traitlets import Dict, Unicode, List, default, validate
+from traitlets import Bool, Dict, Unicode, List, default, validate

 from jupyterhub.crypto import decrypt, EncryptionUnavailable, InvalidToken
 from jupyterhub.auth import LocalAuthenticator
@@ -31,11 +31,17 @@ def check_user_in_groups(member_groups, allowed_groups):
 class GoogleLoginHandler(OAuthLoginHandler):
     """See https://developers.google.com/identity/protocols/oauth2 for general information."""

+    retrieve_all_scopes_at_once = Bool(
+        os.environ.get('GOOGLE_LOGIN_HANDLER_ALL_SCOPES', 'False').lower() in {'true', '1'},
+        help="Set extra_params `access_type` to `offline` and `approval_prompt` to `force`"
+    ).tag(config=True)
+
     def authorize_redirect(self, *args, **kwargs):
         """Add `access_type`, `approval_prompt` to redirect params"""
-        extra_params = kwargs.setdefault('extra_params', {})
-        extra_params["access_type"] = 'offline'
-        extra_params["approval_prompt"] = 'force'
+        if self.retrieve_all_scopes_at_once:
+            extra_params = kwargs.setdefault('extra_params', {})
+            extra_params["access_type"] = 'offline'
+            extra_params["approval_prompt"] = 'force'

         return super().authorize_redirect(*args, **kwargs)
```

Reply:
I don't have a good overview of how people currently/typically configure the Google authenticator. But opt-in sounds like a good way to reduce the chances of breaking stuff for people who are currently happily running the authenticator

Reply:
I guess it's a balance between configurability and code complexity.

Is it easy for you to post some screenshots of the behaviour with and without this PR?

Reply:
@manics the screenshots are the same with and without, the difference is that without these I do not get a refresh token and with them I do

Reply:
logs with:
```
[D 2020-05-05 08:28:29.910 JupyterHub google:41] retrieve_all_scopes_at_once is set to True
[I 2020-05-05 08:28:29.911 JupyterHub log:174] 302 GET /hub/oauth_login?next= -> https://accounts.google.com/o/oauth2/v2/auth?response_type=code&redirect_uri=https%3A%2F%2FREDACTED%3A8000%2Fhub%2Foauth_callback&client_id=REDACTED&state=[secret]&access_type=offline&approval_prompt=force&scope=openid+email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive (@::ffff:127.0.0.1) 1.58ms
[D 2020-05-05 08:30:18.805 JupyterHub google:157] user contains {'access_token': 'REDACTED', 'expires_in': 3599, 'refresh_token': 'REDACTED', 'scope': 'https://www.googleapis.com/auth/userinfo.email openid https://www.googleapis.com/auth/drive', 'token_type': 'Bearer', 'id_token': 'REDACTED'}
[D 2020-05-05 08:30:20.057 JupyterHub google:222] user_info contains {'name': 'person', 'auth_state': {'access_token': 'REDACTED', 'refresh_token': 'REDACTED', 'google_user': {'id': 'REDACTED', 'email': 'REDACTED', 'verified_email': True, 'picture': 'REDACTED', 'hd': 'example.com', 'google_groups': ['agroup', 'anothergroup', 'andanotherone']}}, 'admin': True}
```

logs without

```
[D 2020-05-05 08:46:58.945 JupyterHub google:41] retrieve_all_scopes_at_once is set to True
[I 2020-05-05 08:46:58.947 JupyterHub log:174] 302 GET /hub/oauth_login?next= -> https://accounts.google.com/o/oauth2/v2/auth?response_type=code&redirect_uri=https%3A%2F%2FREDACTED%3A8000%2Fhub%2Foauth_callback&client_id=REDACTED&state=[secret]&scope=openid+email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive (@::ffff:127.0.0.1) 2.13ms
[D 2020-05-05 08:47:16.677 JupyterHub google:157] user contains {'access_token': 'REDACTED', 'expires_in': 3599, 'scope': 'openid https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/drive', 'token_type': 'Bearer', 'id_token': 'REDACTED'}
[D 2020-05-05 08:47:16.727 JupyterHub google:194] Refresh token was empty, will try to pull refresh_token from previous auth_state
[D 2020-05-05 08:47:16.734 JupyterHub google:198] encrypted_auth_state was found, will try to decrypt and pull refresh_token from it
[D 2020-05-05 08:47:17.626 JupyterHub google:222] user_info contains {'name': 'person', 'auth_state': {'access_token': 'REDACTED', 'refresh_token': 'REDACTED', 'google_user': {'id': 'REDACTED', 'email': 'REDACTED', 'verified_email': True, 'picture': 'REDACTED', 'hd': 'example.com', 'google_groups': ['agroup', 'anothergroup', 'andanotherone']}}, 'admin': True}
```

Reply:
The most important part is `user contains` I added this line `self.log.debug("user contains %s", user)` right after this line https://github.com/jupyterhub/oauthenticator/blob/75a383e88073f1e58e1fd996a3271aae2c256f29/oauthenticator/google.py#L136

Reply:
Forgot to add that I revoked access to Jupyterhub each time I tried the steps described above.